### PR TITLE
Add native failed download handling

### PR DESCRIPTION
**Pull request by medariox**
- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

#### TODO:
- [x] Standardize clients, with a status object.
- [x] Add postprocessing queue / queueitems
- [x] Tidy up the postprocessing queue. (add to server/status, add option to manually force a queued PP)
- [x] Start PP on completed download
  - [x] Add properties `destination` (file_path) to the torrent clients
    - [x] nzbget
    - [x] sabnzbd
    - [x] qbittorrent
    - [x] rtorrent
    - [x] transmission
    - [x] deluge
- [x] Refactor PostProcess.process(), Not much of a refactor was needed.
- [x] Update db, when PP is succesfull
- [x] Add UI settings for FDH
- [x] Implement failed download handling
- [x] Add postprocessing queue to home/status
- [ ] Pause a postprocess queue item from UI (out of scope for this PR)
- [x] Add options for:
  - [x] global torrent ratio
  - [x] torrent ratio limit per provider
  - [x] Remove torrent after postprocessing
  - [x] Remove torrent after postprocessing AND seeding finishes
  - [x] Remove torrent after postprocessing with data (only for copy)
  - [x] Pause torrent after processing
  - [x] Pause torrent after processing and seeding finishes

#### Out of scope
  - Re-order pp queue items
  - Cancel pp queue items
  - Create a component with a view on all torrent's and nzb downloading with their progress

### CODE REVIEW

**Comment by p0psicles** [comment]
```yml
---
version: "2.1"
services:
  nzbget:
    image: ghcr.io/linuxserver/nzbget
    container_name: nzbget
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/Amsterdam
    volumes:
      - nzbget-data:/config
      - downloads:/downloads
    ports:
      - 6789:6789
    restart: unless-stopped
    networks:
      - media
  rtorrent:
    image: linuxserver/rutorrent
    container_name: rtorrent
    restart: always
    networks:
      - media        
    ports:
      - 5000:5000
      - 6881:6881/udp
      - 8180:80
      - 51413:51413
    volumes:
      - torrent-data:/config
      - downloads:/downloads
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/Amsterdam
  sabnzbd:
    image: linuxserver/sabnzbd
    container_name: sabnzbd
    restart: always
    networks:
      - media        
    ports:
      - 8008:8080
      - 9009:9090
    volumes:
      - nzbdata:/config
      - complete:/downloads
      - incomplete:/incomplete-downloads
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/Amsterdam
  transmission:
    image: linuxserver/transmission
    container_name: transmission
    restart: always
    networks:
        - media
    ports:
        - 9091:9091
        - 51414:51413
        - 51414:51413/udp
    volumes:
        - transmission-config:/config
        - downloads:/downloads
        - watch:/watch
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/Amsterdam
  qbittorrent:
    image: ghcr.io/linuxserver/qbittorrent
    container_name: qbittorrent
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/Amsterdam
      - UMASK_SET=022
      - WEBUI_PORT=8090
    volumes:
      - qbittorrent-config:/config
      - downloads:/downloads
    networks:
      - media
    ports:
      - 6882:6881
      - 6882:6881/udp
      - 8090:8090
    restart: unless-stopped

volumes:
  nzbget-data:
  downloads:
  torrent-data:
  nzbdata:
  complete:
  incomplete:
  transmission-config:
  qbittorrent-config:
  watch:

networks:
  media:
    external: true
```

**Comment by p0psicles** [comment]
@medariox we should think about what are worthwhile and practical statusses to store.
This is what I have now. (have done nzbget and sab). So I still need to get more experience with the different torrent status.
Maybe you can provide some input for the torrents.

    SNATCHED = 0  # 0
    PAUSED = 1  # 1
    DOWNLOADING = 1 << 1  # 2
    DOWNLOADED = 1 << 2  # 4
    SEEDED = 1 << 3  # 8
    # Can trigger a new snatch if configured.
    FAILED = 1 << 4  # 16
    # Failed, but won't trigger a new snatch.
    # Can be manually aborted, or Failed by client because it's a Dupe (see nzbget)
    ABORTED = 1 << 5  # 32
    EXTRACTING = 1 << 6  # 64
    COMPLETED = 1 << 7  # 128
    POSTPROCESSED = 1 << 8  # 256

**Comment by p0psicles** [comment]
Fix #8735

File: medusa/schedulers/utils.py [hunk 1]
```diff
@@ -18,9 +18,9 @@
     ('postProcess', 'Post Process', 'post_processor_scheduler'),
     ('subtitlesFinder', 'Subtitles Finder', 'subtitles_finder_scheduler'),
     ('traktChecker', 'Trakt Checker', 'trakt_checker_scheduler'),
-    ('downloadHandler', 'Torrent Checker', 'download_handler_scheduler'),
```
Comment (medariox) [inline medusa/schedulers/utils.py]: @p0psicles 
I think this was removed by mistake?

Comment (p0psicles) [reply medusa/schedulers/utils.py]: It's there. A few lines up. At least it should be. I'll check

**Comment by XxUnkn0wnxX** [comment]
well with in regards to torrent PP

```
as for torrent removal
1. copy/link: link or copy files without removal of torrent in client
2. move: move the files & do remove & delete function on torrent client
3. copy/link or copy files with removal just in the client
```
not sure how it is for other clients but for qbit: https://github.com/qbittorrent/qBittorrent/wiki/WebUI-API-(qBittorrent-4.1)#delete-torrents

`/api/v2/torrents/delete?hashes=8c212779b4abde7c6bc608063a0d008b7e40ce32&deleteFiles=false` false when doing torrent removal without cleaning up files.
true to delete files & folders when torrent is removed.

**Comment by XxUnkn0wnxX** [comment]
add support for manually added PP? 
well I guess it's out of scope, but say the user were to manually add nzb's or torrents then append the category name, EG medusa to the nzb/torrent when download handler runs it should also process these even though they were not snatched by medusa.

well the very minimum I'd like for @clinton-hall nzbtomedia to use the new background PP that doesn't lock up the UI (Async)

**Comment by clinton-hall** [comment]
@XxUnkn0wnxX 
a new background PP? I'll take a look.

**Comment by p0psicles** [comment]
@clinton-hall the pr is still very much WIP. But what I've created so far is the groudwork to use a queue for pp tasks.

Medusa own manual pp feature makes use of this, and gets updates about the pp process through websockets.

Now I don't know what the best solution is from nzbToMedua POV. But maybe it's  a good idea to have a brainstorm about this. Would you mind joining our discord to talk about it?

**Comment by clinton-hall** [comment]
@p0psicles happy to join discord. just point me in the right direction.

**Comment by p0psicles** [comment]
https://discord.gg/zMdAdUK

**Pull request merged**